### PR TITLE
[query] Improve precision for variance and stddev of equal values

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Improves precision for `variance()` and `stddev()` aggregation operations when the vector contains equal values. In some cases, this caused in mismatch between Prometheus and M3, where the former would return `0.0` and the latter would return a number that is close to zero, but not equal. 

**Special notes for your reviewer**:

Using the same algorithm as in Prometheus

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/query/functions/aggregation/function.go [hunk 1]
```diff
@@ -122,24 +122,39 @@ func stddevFn(values []float64, bucket []int) float64 {
...
+	if len(values) == 0 || len(bucket) == 0 {
 		return math.NaN()
 	}
 
-	average := sum / count
-	sumOfSquares := 0.0
+	// Using Welford's online algorithm for calculating variance
+	// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
+	//
+	// This algorithm is used in Prometheus and also should provide better numerical precision than
```
Comment (vpranckaitis) [inline src/query/functions/aggregation/function.go:132]: Code in prometheus:
https://github.com/prometheus/prometheus/blob/e7e60623ff312ba47906fd4af49a7a24eb32f380/promql/engine.go#L2017-L2021
https://github.com/prometheus/prometheus/blob/e7e60623ff312ba47906fd4af49a7a24eb32f380/promql/engine.go#L2062-L2066

File: src/query/functions/aggregation/function_test.go [hunk 2]
```diff
@@ -161,3 +162,43 @@ func TestAggFns(t *testing.T) {
...
+	values []float64
+}{
+	{
+		"five 1.33e-5",
+		[]float64{1.33e-5, 1.33e-5, 1.33e-5, 1.33e-5, 1.33e-5},
+	},
+	{
+		"three 13.3",
+		[]float64{13.3, 13.3, 13.3},
+	},
```
Comment (vpranckaitis) [inline src/query/functions/aggregation/function_test.go:177]: `13.3` is from the `aggregators.test` file. I've randomly tried a handful of values, `1.33e-5` is one that makes the previous version of `varianceFn()`/`stddevFn()` fail the test.

File: src/query/test/compatibility/testdata/aggregators.test [hunk 3]
```diff
@@ -101,13 +101,11 @@ load 5m
...
-#eval instant at 50m stddev(http_requests)
-#  {} 0.0
+eval instant at 50m stddev(http_requests)
+  {} 0.0
 
-# FAILING issue #11. (it is almost zero)
-#eval instant at 50m stdvar(http_requests)
-#  {} 0.0
+eval instant at 50m stdvar(http_requests)
+  {} 0.0
```
[the remark this answers is not in the document]
Comment (linasm) [reply src/query/test/compatibility/testdata/aggregators.test:108]: This is unexpected. If this was not failing, we would not have noticed the issue.
Please you check, maybe the tolerance was added with some later commit (after these tests were commented out)?
If this tolerance is no longer needed, we might want to abandon it.

File: src/query/functions/aggregation/function.go [hunk 4]
```diff
@@ -122,24 +122,39 @@ func stddevFn(values []float64, bucket []int) float64 {
...
+	// the straight-forward implementation of the variance formula. The algorithm iterates through the values
+	// and at the each step recalculates mean and variance of the values seen so far.
+
+	count := 0
+	partialMean := 0.0
+	partialVarTimesCount := 0.0 // for better precision, calculate `variance * count` and divide at the end
```
Comment (linasm) [inline src/query/functions/aggregation/function.go:138]: Nit: consider putting these into a `var (...)` block. Usually it looks cleaner that way.

**Review by linasm — COMMENTED** [review]
LGTM, just 1 nit. Also a question about the tolerance in tests.

File: src/query/functions/aggregation/function.go (hunk 4 above)
Comment (vpranckaitis) [reply src/query/functions/aggregation/function.go]: https://github.com/m3db/m3/pull/2799/commits/e5a9d981953cde0c2a051bd129a4506dea1f15dc

File: src/query/test/compatibility/testdata/aggregators.test (hunk 3 above)
[the remark this answers is not in the document]
Comment (vpranckaitis) [reply src/query/test/compatibility/testdata/aggregators.test:108]: After discussing with @linasm we've decided to update this line https://github.com/m3db/m3/commit/5db10aabed41db690c96c92d14860d3c26d67f02#diff-25ae4a7d6eea65e4db2787e6a61f5aea2646020856df3c3f5c7a5cb6b368be1cL565-R565

File: src/query/test/compatibility/test.go [hunk 5]
```diff
@@ -562,7 +562,7 @@ func almostEqual(a, b float64) bool {
 	diff := math.Abs(a - b)
 
 	if a == 0 || b == 0 || diff < minNormal {
-		return diff < epsilon
+		return diff < epsilon*minNormal
```
Comment (arnikola) [inline src/query/test/compatibility/test.go:565]: This seems a bit weird, thought that `minNormal` is the lowest possible float, although guess this is what Prom does in their tests too 🤷‍♂️

**Review by arnikola — APPROVED** [review]
LGTM

File: src/query/test/compatibility/test.go (hunk 5 above)
Comment (vpranckaitis) [reply src/query/test/compatibility/test.go:565]: TIL: min-normal is not the lowest one, [a nice description in here](https://programming.guide/normal-vs-subnormal-floats.html). Seems that when you go below that, you start losing precision (number of digits) and there might be [performance penalty](https://en.wikipedia.org/wiki/Subnormal_numbers#Performance_issues).

**Pull request merged**